File: 2493-reverse-odd-levels-of-binary-tree/2493-reverse-odd-levels-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode reverseOddLevels(TreeNode root) {
        dfs(root.left, root.right, 1);
        return root;
    }

    public void dfs(TreeNode lbr, TreeNode rbr, int lvl) {
        if (lbr == null || rbr == null) {
            return;
        }
        if (lvl % 2 == 1) {
            int tmp = lbr.val;
            lbr.val = rbr.val;
            rbr.val = tmp;
        }
        dfs(lbr.left, rbr.right, lvl + 1);
        dfs(lbr.right, rbr.left, lvl + 1);
    }
}

File: 2493-reverse-odd-levels-of-binary-tree/2493-reverse-odd-levels-of-binary-tree.cpp
class Solution {
public:
    TreeNode* reverseOddLevels(TreeNode* root) {
        if (!root) {
            return nullptr;
        }

        queue<TreeNode*> q;
        q.push(root);
        int level = 0;

        while (!q.empty()) {
            int n = q.size();
            vector<TreeNode*> tmp;

            for (int i = 0; i < n; i++) {
                TreeNode* node = q.front();
                q.pop();
                tmp.push_back(node);

                if (node->left) {
                    q.push(node->left);
                }

                if (node->right) {
                    q.push(node->right);
                }
            }

            if (level % 2 == 1) {
                int i = 0, j = tmp.size() - 1;
                while (i < j) {
                    swap(tmp[i]->val, tmp[j]->val);
                    i++;
                    j--;
                }
            }

            level++;
        }

        return root;
    }
};
